ENH: Add option to display slice edges in 3D view

Punzo opened the pull request:
- Implemented functionality to show slice edges in the 3D view.
- Added user interface controls to control the visibility of slice edges.
- Implemented linking controls to synchronize the visibility of slice edges across different views.

![image](https://github.com/user-attachments/assets/eb4afd25-4697-4b62-9028-b935dcac43d2)

Reference: https://discourse.slicer.org/t/show-slice-edges-in-3d-view/11381

Punzo commented:
@lassoan @pieper @jcfr 
https://github.com/user-attachments/assets/fbc93a20-6561-42e6-96b6-f7b926c9bbdc

lassoan commented on Libs/MRML/Core/vtkMRMLSliceNode.h:
```diff
@@ -78,11 +78,27 @@ class VTK_MRML_EXPORT vtkMRMLSliceNode : public vtkMRMLAbstractViewNode
 
+  ///
+  /// The visibility of the slice edge in the 3DViewer.
+  vtkGetMacro(ThreeDSliceEdgeVisibility, int);
```
Using int instead of bool was a limitation in Tcl decades ago. For new Boolean variables we should use bool type.

lassoan replied:
Use `SliceEdgeVisibility3D` instead, it is more similar to `SliceVisible` and this suffix is used elsewhere similarly, too.

lassoan commented on Libs/MRML/Core/vtkMRMLSliceNode.h:
```diff
@@ -78,11 +78,27 @@ class VTK_MRML_EXPORT vtkMRMLSliceNode : public vtkMRMLAbstractViewNode
+
+  ///
+  /// The opacity of the slice edge in the 3DViewer.
+  vtkGetMacro(ThreeDSliceEdgeOpacity, double);
```
Opacity and and radius should not be necessary to tune for each view. It should "just work", the slice view should just look nice. The same scaling factor that is used for relative scaling of markups could be used here (`this->GetScreenScaleFactor() * this->ViewScaleFactorMmPerPixel`). Opacity could be set to 1.0.

lassoan commented on Libs/MRML/DisplayableManager/vtkMRMLThreeDSliceEdgeDisplayableManager.cxx:
```diff
@@ -0,0 +1,455 @@
+#include <vtkColor.h>
+#include <vtkCutter.h>
+#include <vtkFeatureEdges.h>
+#include <vtkImageData.h>
```
remove includes that are no longer needed

lassoan commented on Libs/MRML/DisplayableManager/vtkMRMLThreeDSliceEdgeDisplayableManager.cxx:
```diff
@@ -0,0 +1,455 @@
+#include <cmath>
+
+//---------------------------------------------------------------------------
+class EdgeRepresentation
```
We use the "pipeline" name for such classes.
```suggestion
class SliceEdgePipeline
```

lassoan commented on Libs/MRML/DisplayableManager/vtkMRMLThreeDSliceEdgeDisplayableManager.cxx:
```diff
@@ -0,0 +1,455 @@
+class vtkMRMLThreeDSliceEdgeDisplayableManager::vtkInternal
+{
+public:
+  typedef std::map<vtkMRMLSliceNode*, EdgeRepresentation*> SliceNodesLink;
```
Managing raw pointers is very risky.
```suggestion
  typedef std::map<vtkWeakPointer<vtkMRMLSliceNode>, EdgeRepresentation> SliceNodesLink;
```

lassoan reviewed (commented):
Very nice work, thank you! I've added a few comments inline.

Punzo commented:
> Very nice work, thank you! I've added a few comments inline.

thanks for reviewing! I will apply the review on Monday! have a nice weekend!

lassoan commented:
`GetViewScaleFactorAtPosition()` is somewhat complex operation and it is reimplemented in several widget representation classes already. It would be better not to add yet another copy of this implementation to the slice edge representation but to move the implementation to the MRML camera node and use that everywhere.

Punzo commented:
@lassoan review applied, please have a look again.

NOTE: I moved the method `GetViewScaleFactorAtPosition` in the `vtkMRMLAbstractThreeDViewDisplayableManager` (as a static one) instead of `vtkMRMLCameraNode`. This because the `vtkMRMLInteractionEventData` (needed for the markups representations) can't be included in `vtkMRMLCameraNode` (i.e. core MRML node do not have access to the classes in the displayManagers folder)

jcfr commented:
> at the moment the slice edge display manager is only for display,

:100:  This is great. Thanks for working on this.

> but in future we can improve it and we can add interactions (slice translation, etc...)

Ditto. That would be a sensible next step.

lassoan reviewed (approved):
I've tested and fixed it up a bit more (removing of view nodes, scene close, etc) and added an application setting to set default visibility. It all looks good, ready for integration.

Punzo commented:
> I've tested and fixed it up a bit more (removing of view nodes, scene close, etc) and added an application setting to set default visibility. It all looks good, ready for integration.

thanks Andras for the fixes and merging!
